File: Array/chenjiaming/167/167.c
int* twoSum(int* numbers, int numbersSize, int target, int* returnSize){
    
    int i, j = 0;
    //分配arr数组空间，2位
    int *arr = (int *)malloc(sizeof(int)*2);
    arr[0] = -1;
    arr[1] = -1;
    
    *returnSize = 0;  // 是输出的值的个数
    
    j = numbersSize - 1;

    if (NULL == numbers){return arr;}


    //双指针法
    while (i < j)
    {
        if (numbers[i] + numbers[j] < target)
        {
            i++; //若两者之和小于目标值，需要i往后移使得二者之和变大
        } else if (numbers[i] + numbers[j] > target){
            j--; //若两者之和大于目标值，需要j往前移使得二者之和变小
        } else {
            arr[0] = i + 1;
            arr[1] = j + 1;
            *returnSize = 2;  // 是输出的值的个数  2个
            return arr;
        }
    }

    return arr;
}

File: Array/chenjiaming/167两数之和/167.java
class Solution {
    //自己写的，O(n^2)
    public int[] twoSum1(int[] numbers, int target) {
        for(int i=0;i<numbers.length;i++){
            for(int j=numbers.length-1;j>0;j--){
                if(numbers[i]+numbers[j]==target){
                    return new int[]{++i,++j};
                }
            }
        }
        return null;
    }



//答案，O(n)
    //利用最大最小之和，与 target 做比较，
    //如果sum>target，说明j得小一点，所以j--；
    //如果sum<target，说明i得大一点，所以i++
    public int[] twoSum(int[] numbers, int target) {
        if(numbers == null){
            return null;
        }
        int i = 0;
        int j = numbers.length-1;
        while(i<j){
             int sum = numbers[i]+numbers[j];
             if(sum<target){
                i++;
             }else if(sum>target){
                 j--;
             }else{
                return new int[]{i+1,j+1};
             } 
        }
        return null;
    }

}